### PR TITLE
Reviewed/modernized open() calls

### PR DESCRIPTION
**Pull request by claudep**
By generalizing the `with open...` syntax, we also ensure files are
properly closed.

### CODE REVIEW

File: setup.py [hunk 1]
```diff
@@ -358,17 +358,18 @@ def parse_requirements(file_name):
...
+            # Ignore comments, blank lines and included requirements files
+            if re.match(r'(\s*#)|(\s*$)|(-r .*$)', line):
+                continue
+
+            if re.match(r'\s*-e\s+', line):
+                requirements.append(re.sub(r'\s*-e\s+.*#egg=(.*)$', r'\1', line))
+            elif re.match(r'\s*-f\s+', line):
+                pass
+            else:
+                requirements.append(line.rstrip('\n'))
```
Comment (unho) [inline setup.py]: Is not simpler to drop the `pass` branch?

```python
            if re.match(r'\s*-e\s+', line):
                requirements.append(re.sub(r'\s*-e\s+.*#egg=(.*)$', r'\1', line))
            elif not re.match(r'\s*-f\s+', line):
                requirements.append(line.rstrip('\n'))
```

File: translate/convert/odf2xliff.py [hunk 2]
```diff
@@ -36,11 +36,6 @@
 
 def convertodf(inputfile, outputfile, templates):
     """Convert an ODF package to XLIFF."""
-    # Since the convertoptionsparser will give us an open file, we risk that
-    # it could have been opened in non-binary mode on Windows, and then we'll
-    # have problems, so let's make sure we have what we want.
-    inputfile.close()
-    inputfile = file(inputfile.name, mode='rb')
```
Comment (unho) [inline translate/convert/odf2xliff.py]: What is the reason to remove this given what says the comment?

File: translate/convert/po2wordfast.py [hunk 3]
```diff
@@ -85,9 +85,9 @@ def recursiveprocess(self, options):
         if not options.targetlanguage:
             raise ValueError("You must specify the target language")
         super(WfOptionParser, self).recursiveprocess(options)
-        self.output = open(options.output, 'w')
-        #options.outputarchive.wffile.setsourcelanguage(options.sourcelanguage)
-        self.output.write(str(options.outputarchive.wffile))
+        with open(options.output, 'wb') as self.output:
```
Comment (unho) [inline translate/convert/po2wordfast.py]: Why use `wb` when the previous code was using `w`?

File: translate/misc/optrecurse.py [hunk 4]
```diff
@@ -523,13 +523,13 @@ def openinputfile(self, options, fullinputpath):
...
             return sys.stdin
-        return open(fullinputpath, 'r')
+        return open(fullinputpath, 'rb')
 
     def openoutputfile(self, options, fulloutputpath):
         """Opens the output file."""
         if fulloutputpath is None:
             return sys.stdout
-        return open(fulloutputpath, 'w')
+        return open(fulloutputpath, 'wb')
```
Comment (unho) [inline translate/misc/optrecurse.py]: What is the reason to use `b` in this module opens?

File: translate/storage/cpo.py [hunk 5]
```diff
@@ -717,8 +717,9 @@ def obsolete_workaround():
 
         def writefile(filename):
             self._gpo_memory_file = gpo.po_file_write_v2(self._gpo_memory_file, filename, xerror_handler)
-            with open(filename) as tfile:
-                return tfile.read()
+            with open(filename, 'rb') as tfile:
```
Comment (unho) [inline translate/storage/cpo.py]: Why `rb`?

File: translate/storage/factory.py [hunk 6]
```diff
@@ -63,18 +63,18 @@
 def _examine_txt(storefile):
     """Determine the true filetype for a .txt file"""
     if isinstance(storefile, six.string_types) and os.path.exists(storefile):
-        storefile = open(storefile)
+        storefile = open(storefile, 'rb')
```
Comment (unho) [inline translate/storage/factory.py]: Why `b`?

File: translate/storage/factory.py [hunk 7]
```diff
@@ -63,18 +63,18 @@
...
         raise ValueError("Need to read object to determine type")
     # Some encoding magic for Wordfast
     from translate.storage import wordfast
-    if wordfast.TAB_UTF16 in start.split("\n")[0]:
+    if wordfast.TAB_UTF16 in start.split(b"\n")[0]:
         encoding = 'utf-16'
     else:
         encoding = 'iso-8859-1'
-    start = start.decode(encoding).encode('utf-8')
+    start = start.decode(encoding)
```
Comment (unho) [inline translate/storage/factory.py]: These cleanups seem unrelated to `open`. If they are please put them in a separate commit.

File: translate/storage/mo.py [hunk 8]
```diff
@@ -53,10 +53,9 @@
 
 def mounpack(filename='messages.mo'):
     """Helper to unpack Gettext MO files into a Python string"""
-    f = open(filename)
-    s = f.read()
-    print("\\x%02x" * len(s) % tuple(map(ord, s)))
-    f.close()
+    with open(filename, 'rb') as fh:
```
Comment (unho) [inline translate/storage/mo.py]: Why `b`?

File: translate/storage/project.py [hunk 9]
```diff
@@ -186,7 +186,8 @@ def export_file(self, fname, destfname):
             via the call to
             :meth:`~translate.storage.projstore.ProjectStore.get_file`
             if *fname* is not found in the project."""
-        open(destfname, 'w').write(self.store.get_file(fname).read())
+        with open(destfname, 'wb') as fp:
```
Comment (unho) [inline translate/storage/project.py]: Why `b`?

File: translate/storage/qm.py [hunk 10]
```diff
@@ -73,10 +73,9 @@
 
 def qmunpack(file_='messages.qm'):
     """Helper to unpack Qt .qm files into a Python string"""
-    f = open(file_)
-    s = f.read()
-    print("\\x%02x" * len(s) % tuple(map(ord, s)))
-    f.close()
+    with open(file_, 'rb') as fh:
```
Comment (unho) [inline translate/storage/qm.py]: Why `b`?

File: translate/storage/test_mo.py [hunk 11]
```diff
@@ -156,17 +155,14 @@ def test_output(self):
                 # can skip the checks here.
                 continue
 
-            mo_msgfmt_f = open(MO_MSGFMT)
-            mo_pocompile_f = open(MO_POCOMPILE)
-
-            try:
+            with open(MO_MSGFMT, 'rb') as mo_msgfmt_f:
```
Comment (unho) [inline translate/storage/test_mo.py]: Why `b`?

File: translate/storage/test_mo.py [hunk 12]
```diff
@@ -156,17 +155,14 @@ def test_output(self):
...
                 mo_msgfmt = mo_msgfmt_f.read()
-                print("msgfmt output:")
-                print(repr(mo_msgfmt))
+            print("msgfmt output:")
+            print(repr(mo_msgfmt))
+
+            with open(MO_POCOMPILE, 'rb') as mo_pocompile_f:
```
Comment (unho) [inline translate/storage/test_mo.py]: Why `b`?

File: translate/storage/test_mo.py [hunk 13]
```diff
@@ -156,17 +155,14 @@ def test_output(self):
...
-                print("pocompile output:")
-                print(repr(mo_pocompile))
-                assert mo_msgfmt == mo_pocompile
-            finally:
-                mo_msgfmt_f.close()
-                mo_pocompile_f.close()
+            print("pocompile output:")
+            print(repr(mo_pocompile))
+
+            assert mo_msgfmt == mo_pocompile
```
Comment (unho) [inline translate/storage/test_mo.py]: Please note that this must be only run if the two files have been successfully opened and read.

File: translate/storage/test_mo.py [hunk 14]
```diff
@@ -156,17 +155,14 @@ def test_output(self):
...
+            with open(MO_POCOMPILE, 'rb') as mo_pocompile_f:
                 mo_pocompile = mo_pocompile_f.read()
-                print("pocompile output:")
-                print(repr(mo_pocompile))
-                assert mo_msgfmt == mo_pocompile
-            finally:
-                mo_msgfmt_f.close()
-                mo_pocompile_f.close()
+            print("pocompile output:")
+            print(repr(mo_pocompile))
```
Comment (unho) [inline translate/storage/test_mo.py]: Please put within the `with` block.

File: translate/storage/test_mo.py [hunk 15]
```diff
@@ -156,17 +155,14 @@ def test_output(self):
...
-            mo_msgfmt_f = open(MO_MSGFMT)
-            mo_pocompile_f = open(MO_POCOMPILE)
-
-            try:
+            with open(MO_MSGFMT, 'rb') as mo_msgfmt_f:
                 mo_msgfmt = mo_msgfmt_f.read()
-                print("msgfmt output:")
-                print(repr(mo_msgfmt))
+            print("msgfmt output:")
+            print(repr(mo_msgfmt))
```
Comment (unho) [inline translate/storage/test_mo.py]: Please put within the `with` block.

File: translate/tools/poconflicts.py [hunk 16]
```diff
@@ -180,7 +180,8 @@ def str_len(x):
             for target, unit, filename in translations:
                 unit.othercomments.append("# (poconflicts) %s\n" % filename)
                 conflictfile.units.append(unit)
-            open(fulloutputpath, "w").write(str(conflictfile))
+            with open(fulloutputpath, "wb") as fh:
```
Comment (unho) [inline translate/tools/poconflicts.py]: Why `b`?

File: translate/tools/porestructure.py [hunk 17]
```diff
@@ -101,13 +101,13 @@ def processfile(self, options, fullinputpath):
                     self.checkoutputsubdir(options, os.path.dirname(outputpath))
                     fulloutputpath = os.path.join(options.output, outputpath)
                     if os.path.isfile(fulloutputpath):
-                        outputfile = open(fulloutputpath, 'r')
+                        outputfile = open(fulloutputpath, 'rb')
```
Comment (unho) [inline translate/tools/porestructure.py]: Why `b`?

File: translate/tools/porestructure.py [hunk 18]
```diff
@@ -101,13 +101,13 @@ def processfile(self, options, fullinputpath):
...
                         outputpofile = po.pofile(outputfile)
                     else:
                         outputpofile = po.pofile()
                     outputpofile.units.append(pounit)   # TODO:perhaps check to see if it's already there...
-                    outputfile = open(fulloutputpath, 'w')
-                    outputfile.write(str(outputpofile))
+                    with open(fulloutputpath, 'wb') as fh:
```
Comment (unho) [inline translate/tools/porestructure.py]: Why `b`?

File: translate/tools/poterminology.py [hunk 19]
```diff
@@ -450,7 +449,8 @@ def outputterminology(self, options):
         termitems = self.extractor.filter_terms(terms, nonstopmin=options.nonstopmin, sortorders=options.sortorders)
         for count, unit in termitems:
             termfile.units.append(unit)
-        open(options.output, "w").write(str(termfile))
+        with open(options.output, "wb") as fh:
```
Comment (unho) [inline translate/tools/poterminology.py]: Why `b`?

**Comment by claudep** [comment]
I owe you an explanation about widen usage of the binary mode. With Python 2, adding a 'b' mode to opening/writing files has little effect, except on Windows. With Python 3, the default modes ('r', 'w') are a text mode where Python tries do encode/decode using a default encoding (locale.getpreferredencoding()). Read https://docs.python.org/3.4/library/functions.html#open. I don't think that's what we want with ttk. At several places, we are sniffing the files to detect the encoding, so we can decode the strings using the proper encoding.

**Comment by unho** [comment]
> I owe you an explanation about widen usage of the binary mode. With Python 2, adding a 'b' mode to opening/writing files has little effect, except on Windows. With Python 3, the default modes ('r', 'w') are a text mode where Python tries do encode/decode using a default encoding (locale.getpreferredencoding()). Read https://docs.python.org/3.4/library/functions.html#open. I don't think that's what we want with ttk. At several places, we are sniffing the files to detect the encoding, so we can decode the strings using the proper encoding.

Tests are passing which is a good sign. But obviously we are only testing on Linux with Python 2 and I am worried about if TTK would be work OK on Windows with these changes. It is not a big concern for me since I had trouble setting up TTK on Windows and I don't think this would impact a big number of users.

I am more worried about changing behavior for Python 3, but since that is our final goal I presume you are testing with Python 3 and it works. If it does I am OK with merging the PR provided that other issues are addressed.

There are other not related to `b` usage to be addressed in this PR, so please do.

File: setup.py (hunk 1 above)
Comment (claudep) [inline setup.py]: Technically yes, but I think the current code is a bit simpler to read. But if you prefer this syntax, just confirm you'd like this to be changed.

File: translate/convert/odf2xliff.py (hunk 2 above)
Comment (claudep) [inline translate/convert/odf2xliff.py]: As noted elsewhere in the patch, we are now systematically opening input files in 'rb' mode.

File: translate/storage/factory.py (hunk 7 above)
Comment (claudep) [inline translate/storage/factory.py]: ok, I will delay this change to a further commit.

File: translate/storage/test_mo.py (hunk 15 above)
Comment (claudep) [inline translate/storage/test_mo.py]: I don't understand why. IMHO, the file can be closed as soon as the content have been read.

File: translate/storage/test_mo.py (hunk 13 above)
Comment (claudep) [inline translate/storage/test_mo.py]: Sure, but I don't see what is different than before at this regard.

File: setup.py (hunk 1 above)
Comment (unho) [inline setup.py]: I prefer this to be changed. That `pass` branch does not help understanding the code in any way.

File: translate/storage/test_mo.py (hunk 15 above)
Comment (unho) [inline translate/storage/test_mo.py]: What if the file is not opened at all? We would be trying to output something that wasn't parsed at all.

File: translate/storage/test_mo.py (hunk 13 above)
Comment (unho) [inline translate/storage/test_mo.py]: You are right.

File: translate/storage/test_mo.py (hunk 15 above)
Comment (unho) [inline translate/storage/test_mo.py]: Sorry, you are right. Somehow I was mixing different concepts on my head. My apologies.

File: translate/convert/odf2xliff.py (hunk 2 above)
Comment (unho) [inline translate/convert/odf2xliff.py]: ok

**Comment by unho** [comment]
Will merge after last concerns are addressed and branch is rebased.

**Comment by claudep** [comment]
Rebased.

**Comment by unho** [comment]
Merged. Thanks!
